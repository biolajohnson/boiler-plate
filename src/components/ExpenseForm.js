import React from "react";
import moment from "moment";
import { SingleDatePicker } from "react-dates";
//import 'react-dates/initialize'

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      description: props.expense ? props.expense.description : "",
      notes: props.expense ? props.expense.note : "",
      amount: props.expense ? (props.expense.amount / 100).toString() : "",
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calenderFocused: false,
      error: "",
    };
  }

  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };
  onNotesChange = (e) => {
    const notes = e.target.value;
    this.setState(() => ({ notes }));
  };
  onAmountChange = (e) => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/))
      this.setState(() => ({ amount }));
  };
  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };
  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calenderFocused: focused }));
  };
  onSubmit = (e) => {
    e.preventDefault();
    if (!this.state.description || !this.state.amount) {
      //set error
      this.setState(() => ({ error: "Please insert values!" }));
    } else {
      //clear error
      this.setState(() => ({ error: "" }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note,
      });
    }
  };
  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>
          <input
            placeholder="Description"
            autoFocus
            type="text"
            value={this.state.description}
            onChange={this.onDescriptionChange}
          />
          <input
            placeholder="Amount"
            type="text"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calenderFocused}
            onFocusChange={this.onFocusChange}
            isOutsideRange={() => false}
            numberOfMonths={1}
          />
          <textarea
            placeholder="Add your notes here (optional)"
            value={this.state.note}
            onChange={this.onNotesChange}
          ></textarea>
          <button>Add Expense</button>
        </form>
      </div>
    );
  }
}
