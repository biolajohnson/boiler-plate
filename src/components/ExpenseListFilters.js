import React from 'react'
import { connect } from 'react-redux'
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters'
import { DateRangePicker } from 'react-dates'
import 'react-dates/initialize'

export class ExpenseListFilters extends React.Component{
        state = {
            calenderFocused: null
        }
    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate)
        this.props.setEndDate(endDate)
    }
    onFocusChange = (calenderFocused) => {
        this.setState(() => ({ calenderFocused }))
    }
    onTextChange = (e) => { 
            this.props.setTextFilter(e.target.value)
    }
    onSortChange = (e) => {
        e.target.value === 'date' ? this.props.sortByDate() : this.props.sortByAmount()
    }
    render(){
        return (
            <div>
                <input type="text"
                    value={this.props.filter.text}
                    onChange={this.onTextChange} />
              <select 
              value={this.props.filter.sortBy} 
              onChange={this.onSortChange}>
                <option value="date">Date</option>
                <option value="amount">Amount</option>
              </select>
              <DateRangePicker
               startDate={this.props.filter.startDate} 
               endDate={this.props.filter.endDate}
               onDatesChange={this.onDatesChange}
               focusedInput={this.state.calenderFocused}
               onFocusChange={this.onFocusChange}
               numberOfMonths={1}
               isOutsideRange={() => false}
               startDateId={"expensify-start-date"}
               endDateId={"expensify-end-date"}
               showClearDates={true}
              />
            </div>
        )
    }
}
const mapStateToProps = (state) => {
    return {
        filters: state.filter
    }
}
const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate)),
    sortByAmount: () => dispatch(sortByAmount()),
    sortByDate: () => dispatch(sortByDate())
})

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters)