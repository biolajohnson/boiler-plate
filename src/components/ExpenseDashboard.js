import React from 'react'
import ExpenseList from './ExpenseList'
import ExpenseListFilters from './ExpenseListFilters'

const ExpenseDashboardPage = () => (
    <div>
   Expense Dashboard!
   <ExpenseListFilters />
   <ExpenseList />
    </div>
)
export default ExpenseDashboardPage