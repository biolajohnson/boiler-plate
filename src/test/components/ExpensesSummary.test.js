import React from 'react'
import { shallow } from 'enzyme'
import { ExpensesSummary } from '../../components/ExpensesSummary'


test('should render expenses summary for singular count', () => {
    const wrapper = shallow(<ExpensesSummary expenseCount={1} expensesTotal={2345} />)
    expect(wrapper).toMatchSnapshot()
})

test('should render expenses summary for multiple expenses', () => {
    const wrapper = shallow(<ExpensesSummary expenseCount={34} expensesTotal={23457878773} />)
    expect(wrapper).toMatchSnapshot()
})