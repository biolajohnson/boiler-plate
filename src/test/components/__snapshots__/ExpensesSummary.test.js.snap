// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expenses summary for multiple expenses 1`] = `
<div>
  <h1>
    Viewing 
    34
     
    expenses
     totaling 
    $234,578,787.73
  </h1>
</div>
`;

exports[`should render expenses summary for singular count 1`] = `
<div>
  <h1>
    Viewing 
    1
     
    expense
     totaling 
    $23.45
  </h1>
</div>
`;
